Handle One Account Google SignIn

aweng98 opened the pull request:
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

aweng98 committed:
one account signin

aweng98 commented:
jenkins retest

aweng98 commented:
jenkins retest

aweng98 commented:
![signinspec_oneaccountsignin](https://user-images.githubusercontent.com/35533885/41490534-89210ef2-70c2-11e8-959e-5140e888f965.png)

aweng98 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala:
```diff
@@ -11,24 +11,16 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
-      "should be able to log in and out multiple times as multiple users" in {
-        val Seq(user1, user2) = UserPool.chooseStudents(2)
+      "should be able to log in and out" in {
+        // "harry": "harry.potter@quality.firecloud.org"
```
restrict Google real SignIn to use just one user. It's painful to use x11 forwarding for Google SignIn security question.

breilly2 replied:
We need 2 users, otherwise we lose coverage of the behavior of making sure sign-out followed by sign-in works correctly. The chance that this will fail is low but the cost if it did could be very high!

aweng98 commented:
sorry, I changed it because it was what Jackie wanted on Friday.
xforwarding was too painful for her to do for all test users in user pool.
It is changed to use 2 users (harry and ron) in PR. and added check to
ensure sign out really worked.

On Mon, Jun 18, 2018 at 9:30 AM Brian Reilly <notifications@github.com>
wrote:

> *@breilly2* commented on this pull request.
> ------------------------------
>
> In
> automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala
> <https://github.com/broadinstitute/firecloud-ui/pull/1321#discussion_r196077196>
> :
>
> >
> -      "should be able to log in and out multiple times as multiple users" in {
> -        val Seq(user1, user2) = UserPool.chooseStudents(2)
> +      "should be able to log in and out" in {
> +        // "harry": "harry.potter@quality.firecloud.org"
>
> We need 2 users, otherwise we lose coverage of the behavior of making sure
> sign-out followed by sign-in works correctly. The chance that this will
> fail is low but the cost if it did could be very high!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/broadinstitute/firecloud-ui/pull/1321#discussion_r196077196>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ah40PR9GmaQ4oTv8CuYZ3kfRnSx14zogks5t96tcgaJpZM4UonFk>
> .
>

jacmrob commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala:
```diff
@@ -13,14 +14,27 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
-          }
+    "should be able to log in and out as harry.potter" in {
+      // config is harded to get "harry": "harry.potter@quality.firecloud.org"
+      val harry = Config.Users.testUser
```
let's not call him "harry" here, since that value can be set in the application.conf so it's not necessarily hard-coded to harry.  let's just say 'signinUser1' or something

jacmrob replied:
also since @jmthibault79's changes i'm not sure Config.Users.testUser even exists anymore-- look in FireCloudConfig.scala to confirm

jacmrob commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala:
```diff
@@ -13,14 +14,27 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
+
+    "should be able to log in and out as ron.weasley" in {
+      // config is harded to get "harry": "harry.potter@quality.firecloud.org"
+      val ron = Config.Users.Students.getUserCredential("ron")
```
same here-- let's add another testUser to the Config -- Config.Users.testUser2, and use that

jacmrob commented:
jenkins retest

breilly2 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala:
```diff
@@ -67,17 +68,17 @@ class SignInPage(val baseUrl: String)(implicit webDriver: WebDriver) extends Fir
-        if (chooseAccountLink) {
-          click on (id("identifierLink"))
-          await visible id("identifierId")
+        find(id("identifierLink")).isDefined match {
```
I don't think the use of `match` is more readable than the `if`, especially since you're not doing anything in the `false` case.

You could try something like this:
```
find(id("identifierLink")) foreach { link =>
  click on link
  await visible id("identifierId")
}
```
It reads a little funny because `find` returns only a single element in an `Option`, but `foreach` is a perfectly valid function of `Option`... like `map` but returning `Unit` instead of transforming the value.

aweng98 replied:
I like this

breilly2 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala:
```diff
@@ -108,6 +115,13 @@ class GoogleSignInPopup(implicit webDriver: WebDriver) extends WebBrowser with W
   }
 
+  private def oneAccountSignIn(email: String, password: String): Unit = {
+    emailField(id("email")).value =(email)
```
Unnecessary parentheses.

breilly2 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala:
```diff
@@ -108,6 +115,13 @@ class GoogleSignInPopup(implicit webDriver: WebDriver) extends WebBrowser with W
+    find(id("next")).get.underlying.submit()
+    pwdField(id("Passwd")).value = password
+    find(id("signIn")).get.underlying.submit()
+  }
```
Do we need a `returnFromSignIn()` here? I don't know for sure that this new sign-in flow is in a new window but my impression from the test failure screenshots I've seen is that it is.

aweng98 replied:
seems true. I added `returnFromSignIn` in `signIn`

breilly2 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/library/DataUseAwareSearchSpec.scala:
```diff
@@ -33,11 +34,11 @@ class DataUseAwareSearchSpec extends FreeSpec with WebBrowserSpec with UserFixtu
         modal.submit()
 
-        modal.isVisible shouldBe false
+        eventually { modal.isVisible shouldBe false }
```
These changes look unrelated to sign-in... but I guess that's okay if they help test reliability. Any guidance for when we should be using `eventually`?

aweng98 replied:
unrelated. impulsive.

breilly2 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala:
```diff
@@ -11,23 +12,28 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
+      }
+    }
+
+    "should be able to log in and out as ron.weasley" in {
```
The original intent of this test was lost somewhere along the way. Signing-in as a 2nd user should be in the same test as the other sign-in. It verifies that the new sign-in is respected immediately after sign-out so it's important that the same browser session is used (inside the same `withWebDriver` block).

aweng98 replied:
done

breilly2 reviewed (changes requested):
Mostly small things but worth fixing.

aweng98 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala:
```diff
@@ -104,8 +110,13 @@ class GoogleSignInPopup(implicit webDriver: WebDriver) extends WebBrowser with W
     pressKeys("\n")
+  }
 
-    returnFromSignIn()
```
moved to `signIn` because it's needed by `oneAccountSignIn` and `normalSignIn`

aweng98 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala:
```diff
@@ -119,15 +130,18 @@ class GoogleSignInPopup(implicit webDriver: WebDriver) extends WebBrowser with W
-      await condition(windowHandles.size == 1)
+    try {
+      await condition (windowHandles.size == 1)
+    } catch {
```
bug in original code. timeout exception thrown if condition is not true.

aweng98 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala:
```diff
@@ -11,24 +12,29 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
-          }
+    "should be able to log in and out as different users in same browser" in {
+      val user1 = Users.Students.getUserCredential("harry")
+      withWebDriver { implicit driver =>
```
same web browser used in 2 login

breilly2 reviewed (approved):
:+1:
